desc/singularities.py

```diff
@@ -0,0 +1,733 @@
+    )
+    if diag:
+        dx = eval_x - src_x
+    else:
+        dx = eval_x[:, None] - src_x[None]
+    K = rpz2xyz_vec(src_data["K_vc"], phi=src_data["zeta"])
+    num = jnp.cross(K, dx, axis=-1)
+    r = safenorm(dx, axis=-1)
+    if diag:
+        r = r[:, None]
+    else:
+        r = r[:, :, None]
+    return 1e-7 * safediv(num, r**3)
+
+
+_kernel_biot_savart.ndim = 3
+_kernel_biot_savart.keys = ["R", "zeta", "Z", "K_vc"]
+
+
+def _kernel_biot_savart_A(eval_data, src_data, diag=False):
+    # K  / |r|
+    src_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([src_data["R"], src_data["zeta"], src_data["Z"]]).T)
+    )
+    eval_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([eval_data["R"], eval_data["zeta"], eval_data["Z"]]).T)
+    )
+    if diag:
+        dx = eval_x - src_x
+    else:
+        dx = eval_x[:, None] - src_x[None]
+    K = rpz2xyz_vec(src_data["K_vc"], phi=src_data["zeta"])
+    r = safenorm(dx, axis=-1)
+    if diag:
+        r = r[:, None]
+    else:
+        r = r[:, :, None]
+    return 1e-7 * safediv(K, r)
+
+
+_kernel_biot_savart_A.ndim = 3
+_kernel_biot_savart_A.keys = ["R", "zeta", "Z", "K_vc"]
+
+
+kernels = {
+    "1_over_r": _kernel_1_over_r,
+    "nr_over_r3": _kernel_nr_over_r3,
+    "biot_savart": _kernel_biot_savart,
+    "biot_savart_A": _kernel_biot_savart_A,
+}
+
+
+def virtual_casing_biot_savart(
+    eval_data, eval_grid, src_data, src_grid, interpolator, loop=True
+):
+    """Evaluate magnetic field on surface due to sheet current on surface.
+
+    The magnetic field due to the plasma current can be written as a Biot-Savart
+    integral over the plasma volume:
+
+    𝐁ᵥ(𝐫) = μ₀/4π ∫ 𝐉(𝐫') × (𝐫 − 𝐫')/|𝐫 − 𝐫'|³ d³𝐫'
+
+    Where 𝐉 is the plasma current density, 𝐫 is a point on the plasma surface, and 𝐫' is
+    a point in the plasma volume.
+
+    This 3D integral can be converted to a 2D integral over the plasma boundary using
+    the virtual casing principle [1]_
+
+    𝐁ᵥ(𝐫) = μ₀/4π ∫ (𝐧' ⋅ 𝐁(𝐫')) (𝐫 − 𝐫')/|𝐫 − 𝐫'|³ d²𝐫'
+            + μ₀/4π ∫ (𝐧' × 𝐁(𝐫') × (𝐫 − 𝐫')/ |𝐫 − 𝐫'|³ d²𝐫'
+            + 𝐁(𝐫)/2
+
+    Where 𝐁 is the total field on the surface and 𝐧' is the outward surface normal.
+    Because the total field is tangent, the first term in the integrand is zero leaving
+
+    𝐁ᵥ(𝐫) = μ₀/4π ∫ K_vc(𝐫') × (𝐫 − 𝐫')/ |𝐫 − 𝐫'|³ d²𝐫' + 𝐁(𝐫)/2
+
+    Where we have defined the virtual casing sheet current K_vc = 𝐧' × 𝐁(𝐫')
+
+    Parameters
+    ----------
+    eval_data : dict
+        Dictionary of data at evaluation points. Keys should be those required by
+        kernel as kernel.keys. Vector data should be in rpz basis.
+    eval_grid : Grid
+        Points where integral transform is to be evaluated (eg unprimed coordinates).
+    src_data : dict
+        Dictionary of data at source points. Keys should be those required by
+        kernel as kernel.keys. Vector data should be in rpz basis.
+    src_grid : LinearGrid
```

Comment:
Any check that the source and eval grids are on a single surface. and that is the same surface? that is required right? we check in the objectives but would be good to have somewhere in this one as well since this is a public function

Reply:
This isn't formally a public function (not listed in the docs page), more like the compute utils that are only for devs. We can't do runtime checking here because this function is jit compiled, so we have to do any checking in the objective, but I can add more to the docstring.